### DISCUSSION TITLE
client.read_holding_registers begins with "response" frame?

### DISCUSSION THREAD

**Discussion by volviq** [pymodbus-dev/pymodbus#2180] [Help] [answered]
Hi, 
I'm trying to connect to a Huawei PV unit, and the basic read holding registers somehow fails. 
I'm comparing a Wireshark dump of the working communication implemented in wlcrs/huawei_solar of homeassistant with the communication of my implementation.
And the most significant difference is the package sequence:
Client IP is 92.168.200.3
server IP (Huawei): 192.168.200.1
wireshark dump of my implementation:
```
"No.","Time","Source","Destination","Protocol","Length","Info"
"10","0.849551219","192.168.200.3","192.168.200.1","TCP","74","54472 → 6607 [SYN] Seq=0 Win=32120 Len=0 MSS=1460 SACK_PERM=1 TSval=204759680 TSecr=0 WS=128"
"11","0.851412196","192.168.200.1","192.168.200.3","TCP","62","6607 → 54472 [SYN, ACK] Seq=0 Ack=1 Win=14563 Len=0 MSS=1460 SACK_PERM=1"
"12","0.851537275","192.168.200.3","192.168.200.1","TCP","60","54472 → 6607 [ACK] Seq=1 Ack=1 Win=32120 Len=0"
"13","0.851833971","192.168.200.3","192.168.200.1","Modbus/TCP","66","Response: Trans:     1; Unit:   0, Func:   3: Read Holding Registers"
```

-> as we see, the "Request" is entirely missing!

homeassistant looks like:
```

"No.","Time","Source","Destination","Protocol","Length","Info"
"1","0.000000000","192.168.200.3","192.168.200.1","TCP","74","50226 → 6607 [SYN] Seq=0 Win=32120 Len=0 MSS=1460 SACK_PERM=1 TSval=204808586 TSecr=0 WS=128"
"2","0.001776836","192.168.200.1","192.168.200.3","TCP","62","6607 → 50226 [SYN, ACK] Seq=0 Ack=1 Win=14563 Len=0 MSS=1460 SACK_PERM=1"
"3","0.001939060","192.168.200.3","192.168.200.1","TCP","60","50226 → 6607 [ACK] Seq=1 Ack=1 Win=32120 Len=0"
"4","1.518450180","192.168.200.3","192.168.200.1","Modbus/TCP","66","   Query: Trans:     1; Unit:   0, Func:   3: Read Holding Registers"
"5","1.544207043","192.168.200.1","192.168.200.3","TCP","60","6607 → 50226 [ACK] Seq=1 Ack=13 Win=14551 Len=0"
"6","1.575604328","192.168.200.1","192.168.200.3","Modbus/TCP","65","Response: Trans:     1; Unit:   0, Func:   3: Read Holding Registers"
"7","1.576087797","192.168.200.3","192.168.200.1","TCP","60","50226 → 6607 [ACK] Seq=13 Ack=12 Win=32109 Len=0"

```
and, as it seems logically intuitive, the actual communication after SYN/ACK is a Query.
And the response comes from the Huawei server.

The implementation is the following:

```

from pymodbus.client import ModbusTcpClient
import sys
host_huawei = "192.168.200.1"
port_huawei = "6607"
client = ModbusTcpClient(host_huawei, port=port_huawei)

client.connect()
result = client.read_holding_registers(43006, 1, slave=0)
if result.isError():
    print("Received error: {}".format(result))
else:
    print("result:")
    print("result:", result.registers)
print("closing")
client.close()
    print("read error")

```

The error log shows:
   
  
```
DEBUG:root:start connect
start connect
ERROR:pymodbus.logging:Connection to (192.168.200.1, 6607) failed: timed out
start read
ERROR:pymodbus.logging:Connection to (192.168.200.1, 6607) failed: timed out
Traceback (most recent call last):
  File "/home/simon/dev/modbus/huawei_meter.py", line 28, in <module>
    result = client.read_holding_registers(43006, 1, slave=0)
  File "/home/simon/dev/modbus/.venv/lib/python3.10/site-packages/pymodbus/client/mixin.py", line 107, in read_holding_registers
    return self.execute(
  File "/home/simon/dev/modbus/.venv/lib/python3.10/site-packages/pymodbus/client/base.py", line 395, in execute
    raise ConnectionException(f"Failed to connect[{self!s}]")
pymodbus.exceptions.ConnectionException: Modbus Error: [Connection] Failed to connect[ModbusTcpClient 192.168.200.1:6607]
(.venv) simon@simonsminipc:~/dev/modbus$ python huawei_meter.py 
DEBUG:root:start connect
start connect
ERROR:pymodbus.logging:Connection to (192.168.200.1, 6607) failed: timed out
start read
ERROR:pymodbus.logging:Connection to (192.168.200.1, 6607) failed: timed out
Traceback (most recent call last):
  File "/home/simon/dev/modbus/huawei_meter.py", line 28, in <module>
    result = client.read_holding_registers(43006, 1, slave=0)
  File "/home/simon/dev/modbus/.venv/lib/python3.10/site-packages/pymodbus/client/mixin.py", line 107, in read_holding_registers
    return self.execute(
  File "/home/simon/dev/modbus/.venv/lib/python3.10/site-packages/pymodbus/client/base.py", line 395, in execute
    raise ConnectionException(f"Failed to connect[{self!s}]")
pymodbus.exceptions.ConnectionException: Modbus Error: [Connection] Failed to connect[ModbusTcpClient 192.168.200.1:6607]

```
(FYI the logging code is not shown in the example code)

I'm using pymodbus-3.6.8

Initially I tried pymodbustcp, but there the wireshark look looks even more strange. Interestingly, pymodbustcp worked just Ok when communicating with the heatpump.
Do you see an obvious mistake in the implementation?

**Comment by janiversen** [comment]
Your device does not accept the connection, so of course nothing else happens. 

If e,g, the HA-modbus is still open, then that is the problem, huawei sun2000 (if that is what you refer to) only accept 1 connection.

**Comment by janiversen** [comment, marked as answer]
pymodbustcp is another project and I have no idea how that works!

I happen to have sun2000 on my production system, and it works without problem. I used to have a sync client as you, but switched to async for performance reasons.

**Reply by volviq** [reply]
I'm aware that the Huawei can handle only one Connection, that's why I need to disable the homeassistant module to Connect with the App for example.
still I do not see a reason why pymodbus starts with a "Response" frame. of course the Huawei would not answer that, it is already a reply.
does the Implementation seem OK?
I will try async as next step, i'm just very surprised what might cause this issue.

**Reply by janiversen** [reply]
I am not sure what you are talking about, I think you mix wireshark notation with modbus notation.

The pymodbus client can only send modbus requests.

Activate the pymodbus debug log, then you see the traffic a lot easier.

**Reply by volviq** [reply]
I only Show wireshark Notation. anyway, somehow the async call Just works for me now, thanks for the hint.
as you seem to Work in homeassistant, a question btw: my smart meter gets stuck apparently, it Constantly Shows the same values, even in the Huawei app. only after a reconnection attempt in the "subdevices" Section it sends valid data again. so it seems there is a Modbus command to reinitialise this.
have you observed this before? i did not find a documented modbus register for such a reinit.

**Reply by janiversen** [reply]
This is the pymodbus project, if you think HA have a bug, then please file a bug there and follow their guidelines.

**Reply by volviq** [reply]
as async works, sync obviously has a bug or is incompatible with the huawei. the Other issue with the meter is inside the Huawei, as the Connection has to be reset.

**Reply by janiversen** [reply]
If you think there is a bug, then please use the pymodbus debug call and provide the debug log, so I can see what actually happens.

I used to use the sync client against my sun2000 and had no problems! now I 7se async.

There might be a bug in the sync version, but the problem is more likely to be a timing problem....sun2000 need a pause between connect and first request.. with async we retry automatically so you might not see if the first request fails.
